### PR TITLE
Do not allocate new bitstring/binary on empty append

### PR DESCRIPTION
**Pull request by josevalim**
Thanks to @michalmuskala for the help in writing this.

### CODE REVIEW

File: erts/emulator/test/bs_construct_SUITE.erl [hunk 1]
```diff
@@ -520,6 +520,14 @@ do_more(Bin, Sz) ->
 do_something() ->
     throw(blurf).
 
+append_empty_is_same(Config) when is_list(Config) ->
```
Comment (garazdawi) [inline erts/emulator/test/bs_construct_SUITE.erl]: Can you add a test for when a writeable binary is appended to?

Comment (josevalim) [reply erts/emulator/test/bs_construct_SUITE.erl]: I would love to except I have no idea how to write it. I didn't even know writable binaries were a thing before this PR. Can you provide a quick snippet? Thanks!

Comment (garazdawi) [reply erts/emulator/test/bs_construct_SUITE.erl]: Maybe take a look at what the copy_writeable_binary test does? i.e.

    WriteableBin = <<(id(<<>>))/binary,0,1,2,3,4,5,6,7>>,

Comment (josevalim) [reply erts/emulator/test/bs_construct_SUITE.erl]: Thanks! I have added an assertion and verified that the assertion fails without the proper fix in the writeable branch of the code.

**Comment by garazdawi** [comment]
This test `ts:run(stdlib, io_proto_SUITE, [batch]).` causes a segfault.

**Comment by josevalim** [comment]
For documentation purposes, here is a minimum test case that reproduces the segfault:

```
unicode_options(Config) when is_list(Config) ->
    PrivDir = proplists:get_value(priv_dir,Config),
    %% A string in both russian and greek characters, which is present
    %% in all the internal test files (but in different formats of course)...
    TestData = [1090,1093,1077,32,1073,1080,1075,32,
		1088,1077,1076,32,1092,1086,1100,32,1093,
		1072,1089,32,1089,1086,1100,32,932,951,949,
		32,946,953,947,32,961,949,948,32,
		963,959,967,32,945,961,949,32,966,959,967,949,963],

    Entry = {{utf16, big},"testdata_utf16_big.dat"},
    BomlessDir = filename:join([PrivDir,"BOMLESSDATA"]),
    file:make_dir(BomlessDir),

    WriteBomlessFile = fun({Enc,File},TData,Dir) ->
			  {ok,F} = file:open(filename:join([Dir,File]), [write,binary,{encoding,Enc}]),
			  io:put_chars(F,TData),
			  file:close(F),
			  ok
		end,

    WriteBomlessFile(Entry, TestData, BomlessDir),

    CannotReadFile = fun({Enc,File},Dir) ->
        erlang:display({Enc, File, Dir}),
        {ok,F} = file:open(filename:join([Dir,File]),[read,binary,{encoding,Enc}]),
        file:read(F,10), %% SEGFAULT HERE
        ok
    end,

    CannotReadFile(Entry, BomlessDir),
    ok.
```

Curiously, the test fails with what look like unrelated changes. For example, if I remove the CannotReadFile anonymous function and "inline" the code, then it passes. If i replace file:read/2 by an io:request, which is equivalent implementation wise, it passes too.

I will continue debugging it further when I get more time later today.

**Comment by josevalim** [comment]
@bjorng / @jhogberg this pull request is similar to PRs #1889 and #1890 in that we attempt to avoid allocations when possible like we did for maps and tuples. However, we learned we can't apply such optimizations to `set_element`, because there are instructions that come after `set_element` that assume `set_element` will make a copy of the tuple. Could it be that we are seeing segmentation faults here because there are instructions that assume a copy of non-writable binaries are made by the `bs_append` instruction? Thanks.

**Comment by garazdawi** [comment]
I got curious and think I've figured out what it is. You need to do a GC check that `extra_words` fits on the heap.

**Comment by josevalim** [comment]
Ooooooh, good catch @garazdawi! That fixed the issue. It would have taken me forever to figure it out. Thank you! Could you please give it another look now?

File: erts/emulator/beam/erl_bits.c [hunk 2]
```diff
@@ -1412,6 +1406,19 @@ erts_bs_append(Process* c_p, Eterm* reg, Uint live, Eterm build_size_term,
 	    }
 	}
 
+	if (build_size_in_bits == 0) {
+	    goto return_bin;
+	}
+
+	/*
+	 * Allocate heap space.
+	 */
```
Comment (josevalim) [inline erts/emulator/beam/erl_bits.c]: I just moved this down so we don't allocate more than we are supposed to in case nothing is being appended.

**Comment by garazdawi** [comment]
I'll put it into test once the Christmas holidays are over.

**Comment by josevalim** [comment]
@sverker code updated as suggested, thanks.

**Comment by sverker** [comment]
I pushed yet another commit where I moved the allocation of heap space back up where it was. We must read 'src_bytes' **after** the gc which may move the binary data.

If nothing is appended then we might have done gc a bit earlier than needed.

**Comment by josevalim** [comment]
Beautiful, thanks!

**Comment by garazdawi** [comment]
merged for release in OTP-22. Thanks!

**Pull request merged**